### PR TITLE
Search: Clicking outside overlay closes overlay

### LINKED ISSUE #18937: Search: Clicking outside the overlay should close the modal
#### Is your feature request related to a problem? Please describe.
With the change in design of the search modal in Jetpack 9.5, there's now a clear visible area outside the modal. A Jetpack Search user reported that they expected the modal to close when clicking outside the modal.

#### Describe the solution you'd like
Clicking outside the modal area closes the modal.

Screen Shot 2021-03-03 at 11 47 06

Related: #17554

### PR DESCRIPTION
**Pull request by jsnmoon**
Fixes #18937.

#### Changes proposed in this Pull Request:
Adds an event handler on the window checking for clicks outside the search overlay.

#### Does this pull request change what data or activity we track or use?
None.

#### Testing instructions:

* Apply this change to a site with a Jetpack Search subscription and Instant Search enabled.
* Perform a site search `e.g. `/?s=page`.
* Click outside the overlay.
* Ensure that the overlay closes as expected.

### CODE REVIEW

File: projects/packages/search/src/instant-search/components/overlay.jsx [hunk 1]
```diff
@@ -11,24 +11,37 @@ import { OVERLAY_CLASS_NAME } from '../lib/constants';
...
 	const closeWithEscape = callOnEscapeKey( closeOverlay );
+	const closeWithOutsideClick = callOnOutsideClick( closeOverlay );
 	useEffect( () => {
 		window.addEventListener( 'keydown', closeWithEscape );
+		window.addEventListener( 'click', closeWithOutsideClick );
 		return () => {
-			// Cleanup after event
+			// Cleanup on component dismount
 			window.removeEventListener( 'keydown', closeWithEscape );
+			window.addEventListener( 'click', closeWithOutsideClick );
```
Comment (kangzj) [inline projects/packages/search/src/instant-search/components/overlay.jsx:42]:
```suggestion
			window.removeEventListener( 'click', closeWithOutsideClick );
```
Am I right?

**Review by kangzj — COMMENTED** [review]
It works as described, good job 👍  One potential typo and suggested a change.

File: projects/packages/search/src/instant-search/components/overlay.jsx (hunk 1 above)
Comment (jsnmoon) [reply projects/packages/search/src/instant-search/components/overlay.jsx]: Good catch!

**Review by kangzj — CHANGES_REQUESTED** [review]
I noticed when you open and close the overlay for several times, it causes the browser to hang. Profiled the process, it seems the click event is taking longer each time.

image

Possibly this is related - we should get the element dynamically instead of storing a static handle even it saves some calculations - the element could be destroyed and re-rendered which is causing objects not released?

**Comment by jsnmoon** [comment]
> I noticed when you open and close the overlay for several times, it causes the browser to hang. Profiled the process, it seems the click event is taking longer each time.
>
> Possibly this is related - we should get the element dynamically instead of storing a static handle even it saves some calculations - the element could be destroyed and re-rendered which is causing objects not released?

Good catch. I'm not able to replicate, but it might have been caused by the `resultsContainer` value never getting garbage collected upon rapid summon/dismissal of the search overlay. I went ahead and tweaked the click event handler logic.

**Review by kangzj — DISMISSED** [review]
Good job it works perfectly and code looks good to me 👍

**Review by annacmc — APPROVED** [review]
Works great for me 👍

**Review by kangzj — APPROVED** [review]

**Comment by jsnmoon** [comment]
Hi @jeherve, just wanted to make sure https://github.com/Automattic/jetpack/pull/22461 is working as expected. In this PR, we're making a minor enhancement to the Search package, resulting in a version bump. This version bump is then propagated to Jetpack's composer.json, which triggers the `Jetpack and packages` section of the required review check.

**Comment by anomiex** [comment]
> (major, minor, or patch)

Major or minor, but not patch as a patch bump would not result in a change the "0.8.x-dev" pseudo-version.

Unfortunately the check works on the file level, so we can't specify that certain changes to Jetpack's `composer.json` are ok while others are not.

**Review by jeherve — APPROVED** [review]

**Pull request merged**